docs/life-of-a-request.md

```diff
@@ -0,0 +1,200 @@
+endpoints, and logged-out endpoints, Zulip uses REST for its API. This
+means that we use:
+
+* POST for creating something new where you don't have a unique ID
+* PUT for creating or completely replacing something for which you have
+  a unique ID
+* DELETE for deleting something
+* PATCH for updating or editing attributes of something.
+* GET to get something (read-only)
+* HEAD to check the existence of something to GET, without getting it;
+  useful to check a link without downloading a potentially large link
+* OPTIONS (handled automatically, see more below)
+* POST for everything else
+
+Of these, PUT, DELETE, HEAD, OPTIONS, and GET are *idempotent*, which
+means that you can send the request multiple times and get the same
+state on the server. POST is not idempotent--if I send a message
+multiple times, I will get multiple message. PATCH is special--it can be
+idempotent, and we like it if you write API endpoints where it is
+idempotent, as much as possible.
+
+### PUT vs POST vs PATCH
+
+The difference between these verbs is subtle at first glance, but
+important.
+
+Let's say you have something (the HTTP RFC calls them "resources") like
+a realm emoji. An emoji has two attributes: a name and a url for the
+emoji image. Let's pretend there's also an emoji alternative name
+attribute.
+
+In this case, we should avoid POST, since we have a unique identifier:
+the name. We don't want to create multiple emoji with the same name--
+otherwise it would be ambiguous which emoji we want to select when
+composing a message.
+
+If we want create the emoji, we should use PUT, since we're replacing
+the entire resource at the unique ID (name). If we wanted to simply
+update some attribute of the emoji, either the image url or the
+alternative name, we should use PATCH. Otherwise, to update the image
+url without clearing the alternative name would require knowing the
+alternative name of the emoji we want to upate. If we wanted to make it
+so that you couldn't update the image url without changing the
+alternative name, or at least knowing it, we could enforce using PUT
+for updates to the realm emoji.
+
+http://restful-api-design.readthedocs.io/en/latest/methods.html can be
+a good read.
+
+### OPTIONS
+
+This HTTP verb will yield the allowed methods 
+OPTIONS https://zulip.tabbott.net/api/v1/users
+
+will yield, as a HTTP header:
+
+`Allow: POST, GET`
+
+You can see this reflected in [zproject/urls.py](https://github.com/zulip/zulip/blob/master/zproject/urls.py):
+
+    url(r'^users$', 'zerver.lib.rest.rest_dispatch',
+        {'GET': 'zerver.views.users.get_members_backend',
+         'POST': 'zerver.views.users.create_user_backend'}),
+
+In this way, the API is partially self-documenting.
+
+### Legacy endpoints are used by the web client
+
+Many of endpoints from our legacy JSON API are written without REST in
+mind. They are still largely used by the web client, and use POST.
```

Comment:
This "largely" feels ambiguous to me. It might be read "the majority of calls to the legacy JSON API are by the web client" or be read "The legacy JSON API is still used extensively by the web client." I think it's the latter, but I'm not sure.